File: callback.c
#include <stdio.h>
#include <stdbool.h>

typedef struct
{
	void (*callback)(int);
	int value;
} Result;

Result get(Result (*result)(int), int input);

Result callback(Result (*callback)(int), int input)
{
	return callback(input);
}

Result callback0(int input)
{
	Result result = {NULL, input + 1};
	return result;
}

Result callback1(int input)
{
	return get(callback0, input + 1);
}

Result callback2(int input)
{
	if (input < 2)
	{
		return get(callback1, input);
	}
	Result result = {NULL, input + 3};
	return result;
}

Result call(int input)
{
	return get(callback2, input);
}

File: callback.js
let promises = {};

function convert(offset) {
	const view = new Uint32Array(memory.buffer, offset, 2);
	var result = {
		callback: view[0],
		value: view[1]
	};
	return result;
}

const callback = (output, fn, input) => {
	wasm.instance.exports.callback(output, fn, input);
	var value = convert(output);
	if (value.callback != 0) {
		value = promises[output].promise;
	} else {
		value = value.value;
		delete promises[output]
	}
	return value;
}

const get = (output, fn, input) => {
	new Uint32Array(memory.buffer, output, 2).set([fn, 0]);
	promises[output] = {
		promise: new Promise((resolve, error) => {
			resolve(input);
		}).then(value => callback(output, fn, value)),
		callback: fn
	};
}

const file = fs.readFileSync('./callback.wasm');
let wasm = await WebAssembly.instantiate(file, { "env": { "get": get, "callback": callback } });
let { allocate, release, memory } = wasm.instance.exports;

export async function call(input) {
	input ||= 0;
	const output = allocate(8);
	wasm.instance.exports.call(output, input);
	release(output);
	return promises[output].promise;
};

export { wasm, promises };